
<P>
  <H1>BSGatlas v1 tracks</H1><BR/>
</P>

<p>
    <b>
BSGatlas: An enhanced annotation integration of the Bacillus subtilis genome
    </b>
    <br/>
Geissler&nbsp;AS,
Anthon&nbsp;C,
Alkan&nbsp;F,
Gonz√°lez-Tortuero&nbsp;E,
Poulsen&nbsp;LD,
Kallehauge&nbsp;TK,
Seemann&nbsp;SE,
Vinther&nbsp;J,
Gorodkin&nbsp;J .
    <i>
      Under review
    </i>
    <br/>
<!--
    [
    <a target="_blank"href=#'>
      PubMed
    </a>
    |
    <a target="_blank"href='#'>
      Paper
    </a>
    ]
-->
</p>

<P>
  <h2>Description of tracks separated by group:</h2>
</P>
<P>
  <h3>BSGatlas</h3><br/>
  <UL>
    <LI>
      <B>Genes:</B> Results after merging various gene annotation resources.
      A click on individual gene shows all meta information related to it
      from the used resources.
    </LI>
    <LI>
      <B>Transcriptional Units</B> The resources
      DBTBS, SubtiWiki, and BsubCyc provide transcriptional units. The collection
      of these are shown in this track. A click on an individual TU reveals
      which of these resources annotate it.
    </LI>
    <LI>
      <B>TSS/terminators/UTRs:</B> We integrated the
      existing TSS and TTS data and associated these with the
      nearest genes in the respective direction of transcription.
      We annotate within the region inbetween with
      5'/3' UTRs. We added based on the Trancriptional Units
      internal UTRs.
    </LI>
    <LI>
      <B>transcripts/isoforms:</B> We inferred from the
      set of transcriptional units and TSS/TTS the full transcripts.
    </LI>
    <LI>
      <B>Operon:</B> Transcripts that share at least one gene, are isoforms
      from the same operon.
      The here shown operons were computed according to that statement.
    </LI>
  </UL>
</P>

<P>
  <H3>External Gene Resources</H3><BR/>
  <UL>
    <LI>
      <B>Unmerged Genes:</B> 
      Composite track that contains all resources that were used in the during
      gene merging. The track controls allows the selection of each resource
      and shows the priority level that were assigned to it.
    </LI>
    <LI>
      <B>Sensetive Rfam:</B> 
      A scan of the Bacillus genome at sensetive levels provided
      additional predictions for Rfam 14.1 families.
      We deemed as to sensetive, and were not explicitly included,
      however, we show them here.
    </LI>
  </UL>
</P>
<P>
  <H3>External Transcription start/termination resources</H3><BR/>
  The resources shown below provide sets of positions of
  transcription start and termination sites, as well as binding
  sites of transcriptional factors and untranslated regions (UTRs) as annotated
  in each resources.
  Each resource is represented by a composite track, which allows selection
  by annotation type.
  <UL>
    <LI><B>BsubCyc</B></LI>
    <LI><B>DBTBS</B></LI>
    <LI><B>Nicolas et. al</B></LI>
    <LI><B>BsubCyc</B></LI>
  </UL>
</P>

<P>
  <H3>Tiling-array signals (not shown by default)</H3>
  This category shows the log2 raw signal of the tiling-array experiments
  conducted by Nicolas <i>et al.</i>. 
  The data was retrieved from GEO (acession GSE27219).
</P>

<P>
  <H3>Lifted probes</H3>
  Information of the probes design for the tiling-array experiments. The intial design was for a a former 
 genome assembly, such that the shown information is lifter over to the genome assembly used in this browser based on a pair-wise genome alignment.
</P>

<P>
  <H3>CRISPR Targets</H3>
  Predictions of CRISPR-Cas9 gRNA targets for the entire genome, filtered for
  specificity. The specificity is computed with the CRISPRspec score,
  considering all potential off-target sites with mismatches up to 8 nts with a
  PAM sequence of NGG, NAG and NGA. Per default, only guides with CRISPRspec
  score larger than 7 are shown. A click on a guide reveals a detailed page of
  its potential off-targets (with mismatches up to 4 nts), ranked by their
  CRISPRoff off-target confidence score. Note that, G-U wobbles are not
  considered as a mismatch in the CRISPRoff off-target confidence score
  computation. Both on-/off targets' locations also show information of
  potentially affected gene/operon targets by listing overlapping annotation in
  the BSGatlas. For all predicted on-target sides, we also show the predicted
  efficiency of the genome-editing, estimated with the Azimuth score. Please
  note that a few guides have potentially multiple on-targets with exact
  sequence matches. These are explicitly highlighted and accompanied by
  information on alternative on-target sites.

</P>
