<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8"/>
<script src="lib/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="lib/bootstrap-3.3.5/css/simplex.min.css" rel="stylesheet" />
<script src="lib/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="lib/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="lib/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="lib/kePrint-0.0.1/kePrint.js"></script>

</head>
<body style="background-color:white;">
<table class="table table-striped table-hover" style="font-size: 14px; margin-left: auto; margin-right: auto;">
<caption style="font-size: initial !important;">BSGatlas-gene-4385</caption>
 <thead>
  <tr>
   <th style="text-align:left;"> meta </th>
   <th style="text-align:left;"> info </th>
  </tr>
 </thead>
<tbody>
  <tr grouplength="2"><td colspan="2" style="background-color: #666; color: #fff;"><strong>1 BSGatlas</strong></td></tr>
<tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Name </td>
   <td style="text-align:left;"> lipL </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Type </td>
   <td style="text-align:left;"> CDS </td>
  </tr>
  <tr grouplength="16"><td colspan="2" style="background-color: #666; color: #fff;"><strong>2 SubtiWiki</strong></td></tr>
<tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Alternative Name </td>
   <td style="text-align:left;"> ipa-90d </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> lipL </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> ywfL </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Category </td>
   <td style="text-align:left;"> SW 2 Metabolism </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> SW 2.6 Additional metabolic pathways </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> SW 2.6.2 Biosynthesis of cofactors </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> SW 2.6.2.7 Biosynthesis and scavenging of lipoic acid </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Description </td>
   <td style="text-align:left;"> N-amidotransferase, transfers lipoic acid from lipoyl-[[protein|58CF62EBA28E77668A0B5319142C8AC20D81CD8A]] to [[protein|C4B6C3EE560C8BD353FEABB1A607C89C20DC8D34]], [[protein|262C9FD20C7A70B6F1FEB57735FA800F38EAB25A]], [[protein|02BA02D10DFB06E51101D8CF76BCF5BED94D7CA2]], and [[protein|2F40086E35FA32136B9A89C530A86D714FE9460C]], and from lipoyl-[[protein|02BA02D10DFB06E51101D8CF76BCF5BED94D7CA2]] to [[protein|262C9FD20C7A70B6F1FEB57735FA800F38EAB25A]] and [[protein|2F40086E35FA32136B9A89C530A86D714FE9460C]] </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Enzyme Classifications </td>
   <td style="text-align:left;"> 2.3.1.204 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Function </td>
   <td style="text-align:left;"> lipoic acid metabolism </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Is essential? </td>
   <td style="text-align:left;"> no </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Isoelectric point </td>
   <td style="text-align:left;"> 5.27 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Locus Tag </td>
   <td style="text-align:left;"> BSU37640 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Molecular weight </td>
   <td style="text-align:left;"> 31.2718 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Name </td>
   <td style="text-align:left;"> lipL </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Product </td>
   <td style="text-align:left;"> protein N-octanoyltransferase </td>
  </tr>
  <tr grouplength="7"><td colspan="2" style="background-color: #666; color: #fff;"><strong>3 RefSeq</strong></td></tr>
<tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Alternative Locus Tag </td>
   <td style="text-align:left;"> BSU37640 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Description </td>
   <td style="text-align:left;"> Evidence 1a: Function from experimental evidencesin the studied strain; PubMedId: 21338420, 21338421;Product type e: enzyme </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Functions </td>
   <td style="text-align:left;"> 4: Biosynthesis of cofactors, prostheticgroups, and carriers </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Locus Tag </td>
   <td style="text-align:left;"> BSU_37640 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Name </td>
   <td style="text-align:left;"> lipL </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Title </td>
   <td style="text-align:left;"> amidotransferase of the octanoyl moiety fromoctanoyl-GcvH to the E2 subunit of pyruvate dehydrogenase </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Type </td>
   <td style="text-align:left;"> CDS </td>
  </tr>
  <tr grouplength="17"><td colspan="2" style="background-color: #666; color: #fff;"><strong>4 BsubCyc</strong></td></tr>
<tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Alternative Name </td>
   <td style="text-align:left;"> ipa-90d </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> ywfL </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Citation </td>
   <td style="text-align:left;"> 429-50 (2016)<br>PubMed: 27074917 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> Cronan JE<br>Assembly of Lipoic Acid on Its Cognate Enzymes: an Extraordinary and Essential Biosynthetic Pathway.<br>Microbiol Mol Biol Rev 80(2) </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Comment </td>
   <td style="text-align:left;"> LipL encodes an amidotransferase that is required for lipoate biosynthesis in <i>B. subtilis</i> |CITS: [21338420][21338421]|.

The <i>lipL</i> gene was cloned with a hexahistidine-tag, expressed in <i>E. coli</i> and purified to homogeneity |CITS: [21338421]|.  The purified protein could bind up to two molecules of either |FRAME: LIPOIC-ACID| or |FRAME: CPD-195|.  These molecules do not bind to lysine residues, as they do in the lipoyl domains of 2-oxo acid dehydrogenases, but to two cysteine residues, located at positions C39 and C150.  Mutation analysis demonstrated that only C150 is essential for activity, and the authors suspect that acylation of C39 may have been an artifact.

The enzyme was shown to catalyze amidotransfer of octanoate from an octanoylated GcvH protein (the lipoate binding component of the glycine cleavage system) to the homologous component of the pyruvate dehydrogenase complex, via an octanoyl-enzyme intermediate.  It can not accept free octanoate as a substrate, and thus GcvH is a required intermediate in the lipoylation of other 2-oxo acid dehydrogenases in <i>B. subtilis</i> |CITS: [21338421]|.

One or more strains containing mutant alleles of this gene can be obtained from the <b>Bacillus Genetic Stock Center</b>.  Click <a href="http://pasture.asc.ohio-state.edu/BGSC/index.cfm?Search=ywfL&amp;Go=Search">here</a> to see the list of available strains at BGSC. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Description </td>
   <td style="text-align:left;"> GcvH:[lipoyl domain] amidotransferase </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Enzyme Classifications </td>
   <td style="text-align:left;"> EC-2.3.1.204 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Gene Ontology </td>
   <td style="text-align:left;"> GO:0003824 catalytic activity </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> GO:0006464 cellular protein modification process </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> GO:0009107 lipoate biosynthetic process </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> GO:0009249 protein lipoylation </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> GO:0016415 octanoyltransferase activity </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> GO:0016740 transferase activity </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> GO:0016746 transferase activity, transferring acyl groups </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Locus Tag </td>
   <td style="text-align:left;"> BSU37640 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Molecular weight </td>
   <td style="text-align:left;"> 31.424 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Name </td>
   <td style="text-align:left;"> lipL </td>
  </tr>
  <tr grouplength="41"><td colspan="2" style="background-color: #666; color: #fff;"><strong>7 Nicolas et al. predictions</strong></td></tr>
<tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Expression neg. correlated with </td>
   <td style="text-align:left;"> BSU01910 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> BSU02240 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> BSU02250 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> BSU23450 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> BSU23460 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> BSU23470 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> new_1780399_1780532 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> new_1915001_1915132_c </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> new_213878_213940 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> new_820648_820823 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Expression pos. correlated with </td>
   <td style="text-align:left;"> BSU03560 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> BSU05910 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> BSU05920 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> BSU08650 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> BSU29805 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> BSU30320 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> BSU37650 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> BSU40650 </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> new_2800811_2800889_c </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> new_3105044_3105139_c </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Highly expressed condition </td>
   <td style="text-align:left;"> (Diami) Cells were grown in LB medium at 37°C. At OD540 of 0.3, the culture were divided into four subcultures and diamide 0.6 mM [Diami], paraquat 0.4 mM [Paraq], H2O2 0.1mM [H2O2] or no oxidative drug [Oxctl] were added to the medium. Samples were taken 10 minutes after addition </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (G135) Purified spores were obtained by growing cells in DSM medium (P. Schaeffer, J. Millet, J. P. Aubert, Proc Natl Acad Sci U S A 54, 704, Sep, 1965) at 37°C for 48 hours after which they were washed ten times in ice cold distilled waterover a period of 5 days. Purified spores were heat activated at 70°C  in Tris 10 mM pH8.4 and germination was initiated by the addition of L-alanine 10 mM (A. Moir, J Bacteriol 146, 1106, Jun, 1981). After incubation for one hour at 37°C, the culture was diluted with an equal volume of 2X LBmedium and germinating cells were harvested at 135, 150 or 180 minutes after addition of L-alanine [G135, G150 and G180]. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (G150) Purified spores were obtained by growing cells in DSM medium (P. Schaeffer, J. Millet, J. P. Aubert, Proc Natl Acad Sci U S A 54, 704, Sep, 1965) at 37°C for 48 hours after which they were washed ten times in ice cold distilled waterover a period of 5 days. Purified spores were heat activated at 70°C  in Tris 10 mM pH8.4 and germination was initiated by the addition of L-alanine 10 mM (A. Moir, J Bacteriol 146, 1106, Jun, 1981). After incubation for one hour at 37°C, the culture was diluted with an equal volume of 2X LBmedium and germinating cells were harvested at 135, 150 or 180 minutes after addition of L-alanine [G135, G150 and G180]. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (G180) Purified spores were obtained by growing cells in DSM medium (P. Schaeffer, J. Millet, J. P. Aubert, Proc Natl Acad Sci U S A 54, 704, Sep, 1965) at 37°C for 48 hours after which they were washed ten times in ice cold distilled waterover a period of 5 days. Purified spores were heat activated at 70°C  in Tris 10 mM pH8.4 and germination was initiated by the addition of L-alanine 10 mM (A. Moir, J Bacteriol 146, 1106, Jun, 1981). After incubation for one hour at 37°C, the culture was diluted with an equal volume of 2X LBmedium and germinating cells were harvested at 135, 150 or 180 minutes after addition of L-alanine [G135, G150 and G180]. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (H2O2) Cells were grown in LB medium at 37°C. At OD540 of 0.3, the culture were divided into four subcultures and diamide 0.6 mM [Diami], paraquat 0.4 mM [Paraq], H2O2 0.1mM [H2O2] or no oxidative drug [Oxctl] were added to the medium. Samples were taken 10 minutes after addition </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (Lbexp) NA </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (LBGexp) Cells were grown in Luria-Bertani medium (Sigma) supplemented with glucose 0.3 % [LBG] at 37°C with  vigorous shaking in flasks. Overnight cultures were diluted 2000-fold in fresh pre-warmed medium and samples were collected during the exponential [exp], transition  [tran] and  stationary [stat] phases of the growth cycle . </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (Mt0) Cells were grown in LB medium at 37°C with vigorous shaking.  An exponentially growing culture (O.D.600 approx. 0.25) was divided: one culture acted as the control [no mitomycin C , M0] while mitomycin was added to the second culture to a final concentration of 40 ng/ml [mitomycin, M40].  Samples were harvested at 0, 45 and 90 minutes after mitomycin addition [t0, t45 and t90]. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (Oxctl) Cells were grown in LB medium at 37°C. At OD540 of 0.3, the culture were divided into four subcultures and diamide 0.6 mM [Diami], paraquat 0.4 mM [Paraq], H2O2 0.1mM [H2O2] or no oxidative drug [Oxctl] were added to the medium. Samples were taken 10 minutes after addition </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (Paraq) Cells were grown in LB medium at 37°C. At OD540 of 0.3, the culture were divided into four subcultures and diamide 0.6 mM [Diami], paraquat 0.4 mM [Paraq], H2O2 0.1mM [H2O2] or no oxidative drug [Oxctl] were added to the medium. Samples were taken 10 minutes after addition </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Lowely expressed condition </td>
   <td style="text-align:left;"> (B36) A fresh colony grown on an LB plate was used to inoculate 10 ml of LB and grown for 10 hoursat 30°C. This culture wasused to inoculate 10 ml of MSgg medium (S.S. Branda et al., J Bacteriol 186, 3970, Jun, 2004) and incubated with vigorous shaking. The cultures in MSgg were diluted to the same extent in 96 wells microtiterplates (5 μl for 1.5 ml of medium) and incubated without shaking at 30°C. Cells from the control cultures were harvested after 24 hours of incubation [BT]. Biofilms were harvested from 96 well plates after incubation for 36 hours [B36] and 60 hours [B60]. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (B60) A fresh colony grown on an LB plate was used to inoculate 10 ml of LB and grown for 10 hoursat 30°C. This culture wasused to inoculate 10 ml of MSgg medium (S.S. Branda et al., J Bacteriol 186, 3970, Jun, 2004) and incubated with vigorous shaking. The cultures in MSgg were diluted to the same extent in 96 wells microtiterplates (5 μl for 1.5 ml of medium) and incubated without shaking at 30°C. Cells from the control cultures were harvested after 24 hours of incubation [BT]. Biofilms were harvested from 96 well plates after incubation for 36 hours [B36] and 60 hours [B60]. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (BT) A fresh colony grown on an LB plate was used to inoculate 10 ml of LB and grown for 10 hoursat 30°C. This culture wasused to inoculate 10 ml of MSgg medium (S.S. Branda et al., J Bacteriol 186, 3970, Jun, 2004) and incubated with vigorous shaking. The cultures in MSgg were diluted to the same extent in 96 wells microtiterplates (5 μl for 1.5 ml of medium) and incubated without shaking at 30°C. Cells from the control cultures were harvested after 24 hours of incubation [BT]. Biofilms were harvested from 96 well plates after incubation for 36 hours [B36] and 60 hours [B60]. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (S4) Cells were grown in CH medium at 37°C and sporulation was induced by resuspension in warm sporulation medium as described by Sterlini and Mandelstam (J. M. Sterlini, J. Mandelstam, Biochem J 113, 29, Jun, 1969). The initiation of sporulation was designated T0, the time of resuspension.  Samples were harvested at hourly intervals for 6 hours [S0 to S6] for the first set of experiments and for 8 hours [S0 to S8] for a  second set of experiments. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (S5) Cells were grown in CH medium at 37°C and sporulation was induced by resuspension in warm sporulation medium as described by Sterlini and Mandelstam (J. M. Sterlini, J. Mandelstam, Biochem J 113, 29, Jun, 1969). The initiation of sporulation was designated T0, the time of resuspension.  Samples were harvested at hourly intervals for 6 hours [S0 to S6] for the first set of experiments and for 8 hours [S0 to S8] for a  second set of experiments. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (S6) Cells were grown in CH medium at 37°C and sporulation was induced by resuspension in warm sporulation medium as described by Sterlini and Mandelstam (J. M. Sterlini, J. Mandelstam, Biochem J 113, 29, Jun, 1969). The initiation of sporulation was designated T0, the time of resuspension.  Samples were harvested at hourly intervals for 6 hours [S0 to S6] for the first set of experiments and for 8 hours [S0 to S8] for a  second set of experiments. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (S8) Cells were grown in CH medium at 37°C and sporulation was induced by resuspension in warm sporulation medium as described by Sterlini and Mandelstam (J. M. Sterlini, J. Mandelstam, Biochem J 113, 29, Jun, 1969). The initiation of sporulation was designated T0, the time of resuspension.  Samples were harvested at hourly intervals for 6 hours [S0 to S6] for the first set of experiments and for 8 hours [S0 to S8] for a  second set of experiments. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (T0.30H) Anon-sporulating B. subtilis strain was grown in a modified M9 medium in batch culture (T. Hardiman, K. Lemuth, M. A. Keller, M. Reuss, M. Siemann-Herzberg, J Biotechnol 132, 359, Dec 1, 2007). Glucose was exhausted when the culture reached an OD600 of approx. 10 and this was designated T0 [T0.0H]. 7 samples were harvested at various times before glucose exhaustion [T-5.40H to T-0.40H] and 10 samples at various times after glucose exhaustion [T0.30H to T5.0H]. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (T2.30H) Anon-sporulating B. subtilis strain was grown in a modified M9 medium in batch culture (T. Hardiman, K. Lemuth, M. A. Keller, M. Reuss, M. Siemann-Herzberg, J Biotechnol 132, 359, Dec 1, 2007). Glucose was exhausted when the culture reached an OD600 of approx. 10 and this was designated T0 [T0.0H]. 7 samples were harvested at various times before glucose exhaustion [T-5.40H to T-0.40H] and 10 samples at various times after glucose exhaustion [T0.30H to T5.0H]. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1">  </td>
   <td style="text-align:left;"> (T4.0H) Anon-sporulating B. subtilis strain was grown in a modified M9 medium in batch culture (T. Hardiman, K. Lemuth, M. A. Keller, M. Reuss, M. Siemann-Herzberg, J Biotechnol 132, 359, Dec 1, 2007). Glucose was exhausted when the culture reached an OD600 of approx. 10 and this was designated T0 [T0.0H]. 7 samples were harvested at various times before glucose exhaustion [T-5.40H to T-0.40H] and 10 samples at various times after glucose exhaustion [T0.30H to T5.0H]. </td>
  </tr>
  <tr>
   <td style="text-align:left; padding-left: 2em;" indentlevel="1"> Name </td>
   <td style="text-align:left;"> lipL </td>
  </tr>
</tbody>
</table>
</body>
</html>
