--- 
title: "BSGatlas: Genetrack"
author:
  - name: Adrian Geissler
    affiliation:
    - &RTH Center for non-coding RNA in Technology and Health,
      Department of Veterinary and Animal Sciences, University of Copenhagen,
      Denmark
  - name: Christian Anthon
    affiliation: *RTH
  - name: Stefan Seemann
    affiliation: *RTH
  - name: Jan Gorodkin
    affiliation:
    - *RTH
    - Corresponding author
date: "`r Sys.time()`"
output:
  BiocStyle::html_document:
    self_contained: TRUE
    toc_float: TRUE
    css: pretty_table.css
---


```{r setup, include = FALSE}
source('00_load.R')

load('01_refseq.rda')
load('01_bsubcyc.rda')
load('01_rfam.rda')
load('01_nicolas.rda')
```

# Gene track

The newest refseq annotation is from `r refseq$date`,
which is based on a major annotation refinment
(Borriss R, et al. Microb Biotechnol. 2018).
We downloaded the annotation in the genbank format
and parsed it with the bioconductor package `genbankr`.
We used simple string matching, to infer more specific 
non-coding RNA types. For only
`r refseq$noncoding %>% filter(type == 'unclear') %>% nrow`
of the
`r nrow(refseq$noncoding)` ncRNA genes it is not directly
clear from the annotaiton itself what they are.
(Maybe these can be refined in the merging)

BsubCyc database
version 38 (release Aug 9, 2017)
has a more systems view representation,
which comprises
`r nrow(bsubcyc$coding)` protein coding and
`r nrow(bsubcyc$noncoding)` non-coding genes and strucutres.
`r nrow(bsubcyc$TSS)` 
experimenatly verifed and nucleotide resolution TSSs,
`r nrow(bsubcyc$terminator)` experimentally verified terminator structures.
The latter two provide clear transcribed boundaries for the some of the
`r nrow(bsubcyc$transunit)` transcriptional units it describes.
As meta-information it provides GO-terms and and indication of transcriptional
relationships. 
In addition, the resource provides detailed metabolic and enzymatic reaction
and pathway overview, which will not be considered here.




The 3016 families of rfam 14.1 were used to create ncRNA and structure annotations
based on conserved stucutre modles, as found by cmsearch v1.1.1.
(TODO: Why a low search space)
The search was conducted at three sensetiviy levels.
The conservative level had a maximal E-value cut-off at 1e-6 and required a
match score of at least the gathering score, a number provided for each
family by rfam.
At the medium level, the score requirement was reduced to half of the gathering
score, and at the senseitive level the E-value requirement was
relaxed to 10.
Afterwards, over-lapping matches, possibly by different rfam family models,
were combined and assigned to the family with the highest score.
(@CA is that correct)
Depending on the sensetivity level, the rfam based cmsearch identified
between 214 and 265 non-coding genes and structures.


The nicolas et al study conducted tiling-array experiments in
`r nrow(nicolas$conditions)` different environmental condition.
Within the resolution limit of the tiling-array, they
identified
`r nrow(nicolas$upshifts)` TSSs and
`r nrow(nicolas$downshifts)` termination sites.
Under consideration of these TSSs and termination sites, they identified
`r nicolas$all.features %>% drop_na %>% nrow` novel transcribed regions.
They attempted to classify these as UTRs or novel transcripts, which seems
to have limitations (more later).
They conducted a literature, and found
`r nicolas$high$their.trusted %>% nrow` of these regions as experimentally
verified in the literature at that time.
In addition, they found
`r nicolas$high$their.lit_review %>% nrow` additional ncRNAs
that weren't found by the tiling array.
Based on the comprehensive environmental conditions, they computed
cluster of (anti-)correlated genes, and predicted
`r nrow(nicolas$predicted.asRNA.targets)`
asRNA-mRNA regulatory pairs.



SubtiWiki downloaded on June 3, 2019